cmd/render/render.go

```diff
@@ -0,0 +1,128 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package render
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/ghodss/yaml"
+	performancev2 "github.com/openshift-kni/performance-addon-operators/api/v2"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/kubeletconfig"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/machineconfig"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/runtimeclass"
+	"github.com/openshift-kni/performance-addon-operators/pkg/controller/performanceprofile/components/tuned"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/klog"
+)
+
+type renderOpts struct {
+	ppInputFilePath string
+	assetsDir       string
+	outputDir       string
+}
+
+type manifestable map[string]interface{}
+
+//NewRenderCommand creates a render command.
+func NewRenderCommand() *cobra.Command {
+	renderOpts := renderOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "render",
+		Short: "Render performance-addon-operator CR manifests",
+		Run: func(cmd *cobra.Command, args []string) {
+
+			if err := renderOpts.Validate(); err != nil {
+				klog.Error(err)
+			}
+
+			if err := renderOpts.Run(); err != nil {
+				klog.Fatal(err)
+			}
+		},
+	}
+
+	renderOpts.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (r *renderOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&r.ppInputFilePath, "performance-profile-file-path", r.ppInputFilePath, "Input path for the Performance Profile manifest.")
+	fs.StringVar(&r.assetsDir, "assets-dir-path", components.AssetsDir, "Input path for the assets directory.")
+	fs.StringVar(&r.outputDir, "output-dir", r.outputDir, "Output path for the rendered manifests.")
+}
+
+func (r *renderOpts) Validate() error {
+	if len(r.ppInputFilePath) == 0 || len(r.outputDir) == 0 {
+		return fmt.Errorf("performance-profile-file-path and output-dir must be specified")
+	}
+	return nil
+}
+
+func (r *renderOpts) Run() error {
```

Comment:
the problem here is we have a fully separate path wrt the online one. Say tomorrow we manage a new object, we will need to remember to add it here. How we can more likely we remember to do this?

Reply:
As for now we have no record for the possible objects that we're creating. it is possible to create a central place like pool/list/file which contains all the objects we have and keep it always updated, and both online/offline paths will follow it (no sure about the implementation details and maybe it deserve a separate PR).
But the same question can also asked here, what if forget to update this central place?

Reply:
indeed the problem is we have two completely separate flows :)

Reply:
for the future:  I think we should evaluate a refactoring/code extraction from  controllers/performanceprofile_controller.go:applyComponents